### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [15]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [16]:
# Take a quick look at the data
purchase.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [3]:
# Initial guess is to use the len() function
players_1 = len(purchase)
# However this assumes that there are no repeat players. 

# How many UNIQUE SN's are there?
players_2 = len(pd.unique(purchase["SN"]))

#print(players_1, players_2)
# OOH! Seems like there was a large descrepency between the two approaches.
# Let's see why:
purchase["SN"].value_counts()
# Looks like players be making multiple purchases. 

# Final Output
# players = pd.DataFrame(data = players_2)
players = pd.DataFrame({"Total Players" : [players_2]})
players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
# Quick look
purchase.describe()

Purchase ID         Age     Item ID       Price
count   780.000000  780.000000  780.000000  780.000000
mean    389.500000   22.714103   92.114103    3.050987
std     225.310896    6.659444   52.775943    1.169549
min       0.000000    7.000000    0.000000    1.000000
25%     194.750000   20.000000   48.000000    1.980000
50%     389.500000   22.000000   93.000000    3.150000
75%     584.250000   25.000000  139.000000    4.080000
max     779.000000   45.000000  183.000000    4.990000

In [5]:
# Let's get some other, perhaps more interesting stats
items = len(pd.unique(purchase["Item ID"]))
price = purchase["Price"].mean()
num_purch = len(purchase)
revenue = purchase["Price"].sum()

purch_stats = pd.DataFrame({"Number of Unique Items": [items], "Average Price": [round(price,2)], 
                           "Total Number of Purchases": [num_purch], "Total Revenue": [revenue]})

purch_stats["Average Price"] = purch_stats["Average Price"].map('${:,.2f}'.format)
purch_stats["Total Revenue"] = purch_stats["Total Revenue"].map('${:,.2f}'.format)
purch_stats

Number of Unique Items Average Price  Total Number of Purchases  \
0                     183         $3.05                        780   

  Total Revenue  
0     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [6]:
# We have this many unique players, each will probably have an associated gender.
playing = pd.unique(purchase["SN"])

# But let's check how many genders we have
gen = purchase.Gender.value_counts()
# We have males, females, and otehr/non-discolosed

# Grouping the dataset by screenname
grouped = purchase.groupby("SN")

# Create a new dataframe (Max just lets me see the information)
unique_sn = pd.DataFrame(grouped.max())

# figure out the percentage of individuals who ID as "male"
# By specifying a value in value_counts, it seems to give me the proportions. Cool.
gen_percent = (unique_sn["Gender"].value_counts("Male")*100) 

# Same idea as above, but getting just the counts
gen_count = unique_sn["Gender"].value_counts()

# compiling the results of gen_p and gen_count into a table
genders = pd.DataFrame({"": ["Male", "Female", "Other/Non-Discolosed"], "Total Count": 
                        [gen_count[0], gen_count[1], gen_count[2]], "Percentage of Players": 
                        [round(int(gen_percent[0]), 2), round(gen_percent[1],2), round(gen_percent[2], 2)]})

genders["Percentage of Players"] = genders["Percentage of Players"].map('{:,.2f}%'.format)

genders.set_index("")

Total Count Percentage of Players
                                                       
Male                          484                84.00%
Female                         81                14.06%
Other/Non-Discolosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
# Create a new dataframe for each gender
males = purchase.loc[purchase["Gender"] =="Male"]
females = purchase.loc[purchase["Gender"] =="Female"]
non_disc = purchase.loc[purchase["Gender"] =="Other / Non-Disclosed"]

# Get number of purchases by gender
# Doesn't matter if someone makes multiple purchases. We count it!
male_p = len(males)
female_p = len(females)
non_disc_p = len(non_disc)
total_p = len(purchase)

# Fun thing I just learned: By formatting, I don't have to round. GAH! This is only somewhat frustrating 
# now that I've pulled my hair out trying to get things to round properly. sigh.
# pp = [round((male_p/total_p)*100,2), round((female_p/total_p)*100,2), round((non_disc_p/total_p)*100,2)]
pp = [male_p/total_p *100, female_p/total_p* 100, non_disc_p/total_p*100]

# Can double check these counts with these two lines:
# purch_counts = purchase["Gender"].value_counts()
# purch_counts

# Total purchase value
male_v = males["Price"].sum()
female_v = females["Price"].sum()
non_disc_v = non_disc["Price"].sum()
total_v = purchase.Price.sum()
vp = [male_v/total_v *100, female_v/total_v *100, non_disc_v/total_v *100]

# Average purchase price
male_avep = male_v/male_p
female_avep = female_v/female_p
non_disc_avep = non_disc_v/non_disc_p

# Average Purchase Total per PERSON. 
male_pp = male_v/gen_count[0]
female_pp = female_v/gen_count[1]
non_disc_pp = non_disc_v/gen_count[2]

# Time to make a table so we can see all the results. WOOT. 
purchase_res = pd.DataFrame({"Gender" : ["Male", "Female", "Other/Non-Discolosed"], 
                             "Number of Purchases" : [male_p, female_p, non_disc_p], 
                             "Percentage of Purchases" : pp,
                             "Total of Purchases" : [male_v, female_v, non_disc_v],
                             "Percentage of Total Price" : vp, 
                             "Average Purchase Price" : [round(male_avep, 2), round(female_avep, 2), 
                                                         round(non_disc_avep, 2)], 
                             "Average Purchase per Player" : [round(male_pp, 2), round(female_pp, 2),
                                                              round(non_disc_pp, 2)]})

purchase_res["Percentage of Purchases"] = purchase_res["Percentage of Purchases"].map('{:,.2f}%'.format)
purchase_res["Percentage of Total Price"] = purchase_res["Percentage of Total Price"].map('{:,.2f}%'.format)
purchase_res["Total of Purchases"] = purchase_res["Total of Purchases"].map('${:,.2f}'.format)
purchase_res["Average Purchase Price"] = purchase_res["Average Purchase Price"].map('${:,.2f}'.format)
purchase_res["Average Purchase per Player"] = purchase_res["Average Purchase per Player"].map('${:,.2f}'.format)

purchase_res.set_index("Gender")

Number of Purchases Percentage of Purchases  \
Gender                                                              
Male                                  652                  83.59%   
Female                                113                  14.49%   
Other/Non-Discolosed                   15                   1.92%   

                     Total of Purchases Percentage of Total Price  \
Gender                                                              
Male                          $1,967.64                    82.68%   
Female                          $361.94                    15.21%   
Other/Non-Discolosed             $50.19                     2.11%   

                     Average Purchase Price Average Purchase per Player  
Gender                                                                   
Male                                  $3.02                       $4.07  
Female                                $3.20                       $4.47  
Other/Non-Discolosed                  $3.35                       $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [8]:
# First we need to find the range of ages
youngest = min(purchase.Age)
oldest = max(purchase.Age)
# print(youngest, oldest)
# So we have people in their single digits and people nearing half a century.
# So lets make bins by increments of 5 years
bins = np.arange(0, 55, 5)
bin_names = ['0-5', '6-10', '11-15', '16-20', '21-25', '26-30','31-35', '36-40', '41-45', '46-50']

# Toss them suckers into their respective bins!
purchase["AgeRange"] = pd.cut(purchase["Age"], bins, labels = bin_names)

# Now create a groupby object thing.
age_group = purchase.groupby(purchase["AgeRange"])

# perform some calculations on that groupby object
ag_total = age_group["Price"].sum()
ag_avg_p = age_group["Price"].mean()
ag_count = age_group["Price"].count()
ag_SN = age_group["SN"].unique()
ag_SNL=[]
ag_percent =[]
ag_avg_pr = []
ag_avg_p = round(ag_avg_p,2)

for i in ag_SN:
    ag_SNL.append(len(ag_SN[i]))
    ag_percent.append(round((len(ag_SN[i])/players_2)*100, 2))
        

The_Age = pd.DataFrame({"Age Group" : bin_names,
                       "Number of Players" : ag_SNL,
                       "Percentage of Players" : ag_percent,
                       "Total Number of Purchases" : ag_count,
                       "Total Purchases" : ag_total,
                       "Average Purchase Price" : ag_avg_p})

The_Age["Average Purchase per Player"] = round(The_Age["Total Purchases"]/The_Age["Number of Players"], 2)
The_Age

Age Group  Number of Players  Percentage of Players  \
AgeRange                                                       
0-5            0-5                  0                   0.00   
6-10          6-10                 24                   4.17   
11-15        11-15                 41                   7.12   
16-20        16-20                150                  26.04   
21-25        21-25                232                  40.28   
26-30        26-30                 59                  10.24   
31-35        31-35                 37                   6.42   
36-40        36-40                 26                   4.51   
41-45        41-45                  7                   1.22   
46-50        46-50                  0                   0.00   

          Total Number of Purchases  Total Purchases  Average Purchase Price  \
AgeRange                                                                       
0-5                               0             0.00                     NaN   
6-10                             32           108.96                    3.40   
11-15                            54           156.60                    2.90   
16-20                           200           621.56                    3.11   
21-25                           325           981.64                    3.02   
26-30                            77           221.42                    2.88   
31-35                            52           155.71                    2.99   
36-40                            33           112.35                    3.40   
41-45                             7            21.53                    3.08   
46-50                             0             0.00                     NaN   

          Average Purchase per Player  
AgeRange                               
0-5                               NaN  
6-10                             4.54  
11-15                            3.82  
16-20                            4.14  
21-25                            4.23  
26-30                            3.75  
31-35                            4.21  
36-40                            4.32  
41-45                            3.08  
46-50                             NaN

In [28]:
# BAH!!! APPARENTLY THERE WERE PRESET BINS!!!!!
# GRUMPH!
bins2 = [min(purchase["Age"])-1,9,14,19,24,29,34,39,max(purchase["Age"])+1] 
bin_names2 = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34','35-39', '40+']

# Toss them suckers into their respective bins!
purchase["AgeRange2"] = pd.cut(purchase["Age"], bins2, labels = bin_names2)

# Now create a groupby object thing.
age_group2 = purchase.groupby(purchase["AgeRange2"])

# perform some calculations on that groupby object
ag_total2 = age_group2["Price"].sum()
ag_avg_p2 = age_group2["Price"].mean()
ag_count2 = age_group2["Price"].count()
ag_SN2 = age_group2["SN"].unique()

ag_SNL2=[]
ag_percent2 =[]
ag_avg_pr2 = []
ag_avg_p2 = round(ag_avg_p2,2)

for i in ag_SN2:
    ag_SNL2.append(len(ag_SN2[i]))
    ag_percent2.append(round((len(ag_SN2[i])/players_2)*100, 2))
        

The_Age2 = pd.DataFrame({"Number of Players" : ag_SNL2,
                       "Percentage of Players" : ag_percent2,
                       "Total Number of Purchases" : ag_count2,
                       "Total Purchases" : ag_total2,
                       "Average Purchase Price" : ag_avg_p2})

The_Age2["Average Purchase per Player"] = round(The_Age2["Total Purchases"]/The_Age2["Number of Players"], 2)

The_Age2["Percentage of Players"]= The_Age2["Percentage of Players"].map('{:,.2f}%'.format)
The_Age2["Total Purchases"]= The_Age2["Total Purchases"].map('${:,.2f}'.format)
The_Age2["Average Purchase Price"]= The_Age2["Average Purchase Price"].map('${:,.2f}'.format)
The_Age2["Average Purchase per Player"]= The_Age2["Average Purchase per Player"].map('${:,.2f}'.format)

The_Age2["Num. Players/Num. Purhcases"]= round(The_Age2["Number of Players"]/ The_Age2["Total Number of Purchases"],2)
The_Age2

Number of Players Percentage of Players  Total Number of Purchases  \
AgeRange2                                                                       
<10                       17                 2.95%                         23   
10-14                     22                 3.82%                         28   
15-19                    107                18.58%                        136   
20-24                    258                44.79%                        365   
25-29                     77                13.37%                        101   
30-34                     52                 9.03%                         73   
35-39                     31                 5.38%                         41   
40+                       12                 2.08%                         13   

          Total Purchases Average Purchase Price Average Purchase per Player  \
AgeRange2                                                                      
<10                $77.13                  $3.35                       $4.54   
10-14              $82.78                  $2.96                       $3.76   
15-19             $412.89                  $3.04                       $3.86   
20-24           $1,114.06                  $3.05                       $4.32   
25-29             $293.00                  $2.90                       $3.81   
30-34             $214.00                  $2.93                       $4.12   
35-39             $147.67                  $3.60                       $4.76   
40+                $38.24                  $2.94                       $3.19   

           Num. Players/Num. Purhcases  
AgeRange2                               
<10                               0.74  
10-14                             0.79  
15-19                             0.79  
20-24                             0.71  
25-29                             0.76  
30-34                             0.71  
35-39                             0.76  
40+                               0.92

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [24]:
# BAH!!! I really need to start reading the next section before starting to do things.
# ENJOY!

The_Age3 = The_Age2.drop(axis = 1, labels = {"Percentage of Players", "Number of Players"})

The_Age3

Total Number of Purchases Total Purchases Average Purchase Price  \
AgeRange2                                                                     
<10                               23          $77.13                  $3.35   
10-14                             28          $82.78                  $2.96   
15-19                            136         $412.89                  $3.04   
20-24                            365       $1,114.06                  $3.05   
25-29                            101         $293.00                  $2.90   
30-34                             73         $214.00                  $2.93   
35-39                             41         $147.67                  $3.60   
40+                               13          $38.24                  $2.94   

          Average Purchase per Player  
AgeRange2                              
<10                             $4.54  
10-14                           $3.76  
15-19                           $3.86  
20-24                           $4.32  
25-29                           $3.81  
30-34                           $4.12  
35-39                           $4.76  
40+                             $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [14]:
sn_spenders = pd.DataFrame(grouped["SN"].max())
sn_spenders["Total"] = grouped["Price"].sum()
sn_spenders["Purchase Count"] = grouped["Purchase ID"].count()
sn_spenders.sort_values(by = "Total", inplace = True, ascending = False)
sn_spenders["Average Purchase"] = "NA"

sn_spenders["Average Purchase"] = round(sn_spenders["Total"]/sn_spenders["Purchase Count"], 2)

Big_Spenders = pd.DataFrame(sn_spenders[[ "Purchase Count", "Average Purchase", "Total"]])
Big_Spenders.rename(columns = {"Average Purchase": "Average Purchase Amount", "Total":"Total Purchases"})

Big_Spenders["Average Purchase"] = Big_Spenders["Average Purchase"].map('${:,.2f}'.format)
Big_Spenders["Total"] = Big_Spenders["Total"].map('${:,.2f}'.format)

Big_Spenders.head()



Purchase Count Average Purchase   Total
SN                                                  
Lisosia93                 5            $3.79  $18.96
Idastidru52               4            $3.86  $15.45
Chamjask73                3            $4.61  $13.83
Iral74                    4            $3.40  $13.62
Iskadarya95               3            $4.37  $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [12]:
items = purchase[["SN", "Item ID", "Item Name", "Price"]]
item_id =items.groupby("Item ID")
# item_desc = items.groupby("Item Name")

# Purchase Count
# item_count = pd.DataFrame(item_id["Item ID"].count())
# item_price = pd.DataFrame(item_id["Price"].min())
item_count = item_id["Item ID"].count()
item_price = item_id["Price"].min()
item_names = item_id["Item Name"].min()
item_total =[]

# Apparently item 36 does not exist. RUDE!
# There is probably a better way too accomplish this, but this is working *shrug*
item_rows = item_id["Item ID"].unique()
for i in item_rows:
    item_rows[i] = int(item_rows[i])

# Calculating some sales amounts
for i in item_rows:
    item_total.append(item_count[i]*item_price[i])

All_Things = pd.DataFrame(item_names)
# All_Things["Item Name"] =item_names
All_Things["Number Sold"] = item_count
All_Things["Price"] = item_price
All_Things["Total Sales"] = item_total

All_Things2 = pd.DataFrame(All_Things)

All_Things.sort_values(by = "Number Sold", inplace = True, ascending = False)

All_Things["Price"] = All_Things["Price"].map('${:,.2f}'.format)
All_Things["Total Sales"] = All_Things["Total Sales"].map('${:,.2f}'.format)
All_Things.head()

Item Name  Number Sold  Price  \
Item ID                                                                     
178      Oathbreaker, Last Hope of the Breaking Storm           12  $4.23   
145                              Fiery Glass Crusader            9  $4.58   
108         Extraction, Quickblade Of Trembling Hands            9  $3.53   
82                                            Nirvana            9  $4.90   
19                      Pursuit, Cudgel of Necromancy            8  $1.02   

        Total Sales  
Item ID              
178          $50.76  
145          $41.22  
108          $31.77  
82           $44.10  
19            $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [13]:
All_Things2.sort_values(by = "Total Sales", inplace = True, ascending = False)

All_Things2["Price"] = All_Things2["Price"].map('${:,.2f}'.format)
All_Things2["Total Sales"] = All_Things2["Total Sales"].map('${:,.2f}'.format)
All_Things2.head()

Item Name  Number Sold  Price  \
Item ID                                                                     
178      Oathbreaker, Last Hope of the Breaking Storm           12  $4.23   
82                                            Nirvana            9  $4.90   
145                              Fiery Glass Crusader            9  $4.58   
92                                       Final Critic            8  $4.88   
103                                    Singed Scalpel            8  $4.35   

        Total Sales  
Item ID              
178          $50.76  
82           $44.10  
145          $41.22  
92           $39.04  
103          $34.80

# Some Data Trends/Observations

* Individuals, ages 20-24, spend the most (gross purchase total) on additional items for this game; however per player, the individuals age 35-39 spend the largest amount, followed by individuals under age 10. 


* While there is some overlap between the most popular and the most profitable items, the 5th most popular item is bringing in a whopping $8.16 in profits. Perhaps they should up that price a bit. Or perhaps it's so popular because it's relatively cheap.


* Males certainly make up more of the players than females, but they seem to make proportionally similar numbers of purchases as seen by % of total purchases and % of total price each gender has made


* The ratio of number of players to number of purchses is greatest in the 40+ age group. The other age groups are relatively similar to one another. This suggests that people 40+ years old make fewer purchases than the other age groups, however this may be driven by the relatively small number of individuals within the age group. 


* Item ID 36 seems to not exist. This caused some issues when looking for the most popular and profitable items, but it seems to have worked out in the end. 